lldb/source/Host/windows/MainLoopWindows.cpp

```diff
@@ -44,36 +37,61 @@ MainLoopWindows::~MainLoopWindows() {
 }
 
 llvm::Expected<size_t> MainLoopWindows::Poll() {
-  std::vector<WSAEVENT> events;
+  std::vector<HANDLE> events;
   events.reserve(m_read_fds.size() + 1);
-  for (auto &[fd, info] : m_read_fds) {
-    int result = WSAEventSelect(fd, info.event, FD_READ | FD_ACCEPT | FD_CLOSE);
-    assert(result == 0);
-    UNUSED_IF_ASSERT_DISABLED(result);
-
-    events.push_back(info.event);
+  for (auto &[fd, fd_info] : m_read_fds) {
+    // short circuit waiting if a handle is already ready.
+    if (fd_info.object_sp->HasReadableData())
+      return events.size();
+    events.push_back(fd);
   }
   events.push_back(m_interrupt_event);
 
-  DWORD result =
-      WSAWaitForMultipleEvents(events.size(), events.data(), FALSE,
-                               ToTimeout(GetNextWakeupTime()), FALSE);
+  while (true) {
+    DWORD timeout = INFINITY;
+    std::optional<lldb_private::MainLoopBase::TimePoint> deadline =
+        GetNextWakeupTime();
+    if (deadline) {
+      // Check how much time is remaining, we may have woken up early for an
+      // unrelated reason on a file descriptor (e.g. a stat was triggered).
+      std::chrono::milliseconds remaining =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              deadline.value() - std::chrono::steady_clock::now());
+      if (remaining.count() <= 0)
+        return events.size() - 1;
+      timeout = remaining.count();
+    }
 
-  for (auto &fd : m_read_fds) {
-    int result = WSAEventSelect(fd.first, WSA_INVALID_EVENT, 0);
-    assert(result == 0);
-    UNUSED_IF_ASSERT_DISABLED(result);
+    DWORD result =
+        WaitForMultipleObjects(events.size(), events.data(), FALSE, timeout);
+
+    // A timeout is treated as a (premature) signalization of the interrupt
+    // event.
+    if (result == WAIT_TIMEOUT)
+      return events.size() - 1;
+
+    if (result == WAIT_FAILED)
+      return llvm::createStringError(llvm::mapLastWindowsError(),
+                                     "WaitForMultipleObjects failed");
+
+    // check if interrupt requested.
+    if (result == WAIT_OBJECT_0 + events.size())
+      return result - WAIT_OBJECT_0;
+
+    // An object may be signaled before data is ready for reading, verify it has
+    // data.
+    if (result >= WAIT_OBJECT_0 &&
+        result < WAIT_OBJECT_0 + (events.size() - 1) &&
+        std::next(m_read_fds.begin(), result - WAIT_OBJECT_0)
+            ->second.object_sp->HasReadableData())
+      return result - WAIT_OBJECT_0;
+
+    // If no handles are actually ready then yield the thread to allow the CPU
+    // to progress.
+    std::this_thread::yield();
```

Comment:
What exactly are the situations when we reach this? I don't mind spinning here once or twice, but I'd hate to see this turn into busy wait...

Reply:
WaitForSingleObject/WaitForMultipleObjects will signal on any change to a file handle, so we can't really differentiate the specific event that caused the change. I've updated the code to use a zero byte read instead and now we shouldn't need to keep trying.